### PR TITLE
Fix a few bugs and improve performance

### PR DESCRIPTION
**Pull request by arybczak**
* Drop support for GHC < 8.8
* Fix a space leak in randomBytesIO.
* Use a buffered generator.
* Remove modulo bias from randomRIO.
* Improve performance of randomString.
* Add support for a pool of generators for less contention.

All in all, improves throughput by several orders of magnitude (!)

### CODE REVIEW

**Comment by arybczak** [comment]
Demonstration of modulo bias:

Before:
```
>>> fmap (map length . group . sort) $ runCryptoRNGT rng $ replicateM 1000000 $ randomR (1,6)
[168539,167759,168155,167136,163667,164744]
```

Now:
```
>>> fmap (map length . group . sort) $ runCryptoRNGT rng $ replicateM 1000000 $ randomR (1,6)
[166814,166365,166626,166631,166911,166653]
```

File: crypto-rng.cabal [hunk 1]
```diff
@@ -14,7 +14,7 @@ maintainer:          Mikhail Glushenkov <mikhail.glushenkov@scrive.com>,
 copyright:           Scrive AB
 category:            Crypto
 build-type:          Simple
-tested-with:         GHC ==8.0.2 || ==8.2.2 || ==8.4.4 || ==8.6.5 || ==8.8.3 || ==8.10.1
+tested-with:         GHC ==8.8.4 || ==8.10.7 || ==9.0.2
```
Comment (jsynacek) [inline crypto-rng.cabal:17]: No 9.2.1 here?

File: src/Crypto/RNG.hs [hunk 2]
```diff
@@ -24,118 +20,122 @@
 -- The access to the RNG state is captured by a class.  By making
 -- instances of this class, client code can enjoy RNG generation from
 -- their own monads.
-module Crypto.RNG (
-  -- * CryproRNG class
+module Crypto.RNG
+  ( -- * CryproRNG class
```
Comment (jsynacek) [inline src/Crypto/RNG.hs]: Typo. s/Crypro/Crypto/. It was in the original comment, too.

File: src/Crypto/RNG.hs [hunk 3]
```diff
@@ -24,118 +20,122 @@
...
+  split = error "split"
+  genWord32 (RNG g) = case genBytes 4 g of
+    Left err       -> error $ "genBytes failed: " ++ show err
+    Right (bs, g') -> (mkWord bs, RNG g')
+  genWord64 (RNG g) = case genBytes 8 g of
+    Left err       -> error $ "genBytes failed: " ++ show err
+    Right (bs, g') -> (mkWord bs, RNG g')
+
+mkWord :: (Bits a, Num a) => ByteString -> a
+mkWord bs = BS.foldl' (\acc w -> shiftL acc 8 .|. fromIntegral w) 0 bs
```
Comment (jsynacek) [inline src/Crypto/RNG.hs:77]: This might be totally off-topic, but this piece of code won't work on big endian architectures, like M1 chips. Do we care?

**Review by jsynacek — CHANGES_REQUESTED** [review]
It would also be awesome to add the "demonstration of modulo bias" from the separate comment into the commit message.

File: src/Crypto/RNG.hs (hunk 3 above)
Comment (marco44) [reply src/Crypto/RNG.hs:77]: AFAIK M1 is little endian, like all ARM processors (they can run in big endian, but almost never are

File: crypto-rng.cabal (hunk 1 above)
Comment (arybczak) [reply crypto-rng.cabal:17]: No, one of the dependencies doesn't compile (yet).

File: src/Crypto/RNG.hs (hunk 3 above)
Comment (arybczak) [reply src/Crypto/RNG.hs:77]: Hmm, but these are random bytes, so their order doesn't really matter. Unless I'm missing something?

Comment (jsynacek) [reply src/Crypto/RNG.hs:77]: Yeah, it probably doesn't matter.

**Comment by jsynacek** [comment]
Looks good.

**Review by jsynacek — APPROVED** [review]

**Pull request merged**